### ISSUE TITLE
🐛 [Firebase messaging] Firebase messaging 9.1.3 - Don't appear notifications in foreground in Android

### ISSUE THREAD

**Issue by csacchetti** [FirebaseExtended/flutterfire#6181]
I updated the Firebase messaging code in my app which has changed a lot with the recent version. A nice thing is the possibility to see the push message also in the foreground, using local notifications. Now everything works for me when the app is closed, when the app is in the background. When the app is in the foreground it works on iOS but not Android.
I followed the instructions. Added custom channel on Manifest.
I'll give you my code. If you tell me where I went wrong I'm grateful.

The Manifest

```
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.carlosacchetti.push_notification_example">
    <!-- io.flutter.app.FlutterApplication is an android.app.Application that
         calls FlutterMain.startInitialization(this); in its onCreate method.
         In most cases you can leave this as-is, but you if you want to provide
         additional functionality it is fine to subclass or reimplement
         FlutterApplication and put your custom class here. -->
    <application
        android:name="io.flutter.app.FlutterApplication"
        android:label="push_notification_example"
        android:icon="@mipmap/ic_launcher">
        <activity
            android:name=".MainActivity"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            <!-- Specifies an Android theme to apply to this Activity as soon as
                 the Android process has started. This theme is visible to the user
                 while the Flutter UI initializes. After that, this theme continues
                 to determine the Window background behind the Flutter UI. -->
            <meta-data
              android:name="io.flutter.embedding.android.NormalTheme"
              android:resource="@style/NormalTheme"
              />
            <!-- Displays an Android View that continues showing the launch screen
                 Drawable until Flutter paints its first frame, then this splash
                 screen fades out. A splash screen is useful to avoid any visual
                 gap between the end of Android's launch screen and the painting of
                 Flutter's first frame. -->
            <meta-data
              android:name="io.flutter.embedding.android.SplashScreenDrawable"
              android:resource="@drawable/launch_background"
              />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>
            <!-- <intent-filter>
                <action android:name="FLUTTER_NOTIFICATION_CLICK" />
                <category android:name="android.intent.category.DEFAULT" /> 
            </intent-filter> -->
        </activity> 
        <!-- Don't delete the meta-data below.
             This is used by the Flutter tool to generate GeneratedPluginRegistrant.java -->
        <meta-data
            android:name="flutterEmbedding"
            android:value="2" />
         <!-- A custom Android Notification Channel to deliver FCM notifications on a non-default channel -->
        <meta-data
            android:name="com.google.firebase.messaging.default_notification_channel_id"
            android:value="high_importance_channel" />
    </application>
</manifest>
```

The "main"

```
Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  // If you're going to use other Firebase services in the background, such as Firestore,
  // make sure you call `initializeApp` before using other Firebase services.
  await Firebase.initializeApp();
  print('Handling a background message ${message.messageId}');
}

/// Create a [AndroidNotificationChannel] for heads up notifications
const AndroidNotificationChannel channel = AndroidNotificationChannel(
  'high_importance_channel', // id
  'High Importance Notifications', // title
  'This channel is used for important notifications.', // description
  importance: Importance.max,
);

/// Initialize the [FlutterLocalNotificationsPlugin] package.
final FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin =
    FlutterLocalNotificationsPlugin();

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp();

  // Set the background messaging handler early on, as a named top-level function
  FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);

  /// Create an Android Notification Channel.
  ///
  /// We use this channel in the `AndroidManifest.xml` file to override the
  /// default FCM channel to enable heads up notifications.
  await flutterLocalNotificationsPlugin
      .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin>()
      ?.createNotificationChannel(channel);

  /// Update the iOS foreground notification presentation options to allow
  /// heads up notifications.
  await FirebaseMessaging.instance.setForegroundNotificationPresentationOptions(
    alert: true,
    badge: true,
    sound: true,
  );

  runApp(MyApp());
}
```

The "HomePage"

```
@override
  void initState() {
    FirebaseMessaging.instance
        .getInitialMessage()
        .then((RemoteMessage message) {
      if (message != null) {
        // Navigator.pushNamed(context, '/message',
        //     arguments: MessageArguments(message, true));
      }
    });
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      RemoteNotification notification = message.notification;
      AndroidNotification android = message.notification?.android;

      if (notification != null && android != null) {
        flutterLocalNotificationsPlugin.show(
            notification.hashCode,
            notification.title,
            notification.body,
            NotificationDetails(
              android: AndroidNotificationDetails(
                channel.id,
                channel.name,
                channel.description,
                // TODO add a proper drawable resource to android, for now using
                //      one that already exists in example app.
                icon: 'launch_background',
              ),
            ));
      }
    });

    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      print('A new onMessageOpenedApp event was published!');
      // Navigator.pushNamed(context, '/message',
      //     arguments: MessageArguments(message, true));
    });

    super.initState();
  }
```

Where is the error that it does not allow foreground notifications in Android?

  [1]: https://i.stack.imgur.com/py1j4.png

**Comment by darshankawar** [comment]
@csacchetti 
Can you provide `flutter doctor -v` and `flutter pub deps -- --style=compact` ?

Can you also try the official plugin repo [example](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example) and see with it if you get same behavior or not ?

**Comment by csacchetti** [comment]
Here is flutter doctor -v

[✓] Flutter (Channel stable, 2.0.6, on macOS 11.2.3 20D91 darwin-x64, locale it-IT)
    • Flutter version 2.0.6 at /Users/carlosacchetti/Developer/flutter
    • Framework revision 1d9032c7e1 (3 weeks ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/carlosacchetti/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2021.1.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.22.0

[✓] Connected device (3 available)
    • sdk gphone x86 (mobile) • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • iPhone 12 Pro (mobile)  • BD691460-BEF6-4934-8147-3364A82CEF14 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)            • chrome                               • web-javascript • Google Chrome 90.0.4430.212

• No issues found!

**Comment by csacchetti** [comment]
Here is flutter pub deps - --style = compact

Dart SDK 2.12.3
Flutter SDK 2.0.6
push_notification_example 1.0.0+1

dependencies:
- cupertino_icons 1.0.0
- firebase_core 1.1.0 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_messaging 9.1.3 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- flutter_local_notifications 5.0.0+4 [flutter flutter_local_notifications_platform_interface platform timezone]

dev dependencies:
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]

transitive dependencies:
- async 2.5.0 [collection]
- boolean_selector 2.1.0 [source_span string_scanner]
- characters 1.1.0
- charcode 1.2.0
- clock 1.1.0
- collection 1.15.0
- fake_async 1.2.0 [clock collection]
- firebase_core_platform_interface 4.0.0 [flutter meta plugin_platform_interface]
- firebase_core_web 1.0.2 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_messaging_platform_interface 2.1.3 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 1.0.6 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- flutter_local_notifications_platform_interface 3.0.0 [flutter plugin_platform_interface]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- js 0.6.3
- matcher 0.12.10 [stack_trace]
- meta 1.3.0
- path 1.8.0
- platform 3.0.0
- plugin_platform_interface 2.0.0 [meta]
- sky_engine 0.0.99
- source_span 1.8.0 [charcode collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- string_scanner 1.1.0 [charcode source_span]
- term_glyph 1.2.0
- test_api 0.2.19 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timezone 0.7.0 [path]
- typed_data 1.3.0 [collection]
- vector_math 2.1.0

**Comment by csacchetti** [comment]
I can't try the official plugin repo example because to send messages I use my account's Firebase console while the plugin sends messages directly.

**Comment by darshankawar** [comment]
@csacchetti 
There have been updates made to the plugin after 9.1.3 and [latest](https://pub.dev/packages/firebase_messaging/install) is 10.0.0. Can you upgrade to latest plugin version and try to see if it helps ?

**Comment by csacchetti** [comment]
I have tried with the latest versions of Firebase messaging and Faribase core but the result is the same. All notifications work apart from the foreground in Android.
The only difference from the example is not to send the notification from the app but from the firebase console. I doubt that there is a parameter to be set in the console that I do not know. On the last screen, before sending the notification, the console asks if you want to indicate an Android notification channel. I put the ID of the one I created specifically for "high_importance_channel" notifications, but I don't know if something else is needed. I attach the image of the last screen.

<img width="991" alt="Schermata 2021-05-21 alle 09 57 21" src="https://user-images.githubusercontent.com/46923349/119103339-56518b00-ba1b-11eb-804f-a53bfcb9613d.png">

**Comment by darshankawar** [comment]
@csacchetti 
What's the payload are you using for the messages ?

When you say `notifications don't appear`, does it show any related errors in the console ?

**Comment by darshankawar** [comment]
May be related / similar to https://github.com/FirebaseExtended/flutterfire/issues/4261

**Comment by csacchetti** [comment]
I tried to use the example app directly and I realized that both with mine and with the example one when they are launched this error occurs which then does not block the app but it could be a clue.

<img width="744" alt="Schermata 2021-05-24 alle 17 24 06" src="https://user-images.githubusercontent.com/46923349/119376003-c7af6900-bcbb-11eb-8931-3ae25747aede.png">

**Comment by csacchetti** [comment]
To answer your questions directly:
- as payload I use the console of the firebase site.
- Apart from that mistake when I started the app I didn't notice any particular errors. When I send the message the app says that
"D / FLTFireMsgReceiver (28106): broadcast received for message"
If I look if the data like "title", "body" etc. they arrived, I see that they arrived regularly. Simply when it is in the foreground, the notification window at the top does not appear.

**Comment by darshankawar** [comment]
Thanks for the details. Keeping this issue open for further insights and investigation from team.

**Comment by tonini46** [comment]
> To answer your questions directly:
> 
> * as payload I use the console of the firebase site.
> * Apart from that mistake when I started the app I didn't notice any particular errors. When I send the message the app says that
>   "D / FLTFireMsgReceiver (28106): broadcast received for message"
>   If I look if the data like "title", "body" etc. they arrived, I see that they arrived regularly. Simply when it is in the foreground, the notification window at the top does not appear.

same here

**Comment by mantishK** [comment]
I was facing the same issue. I noticed that when I reinstalled the app, signed up, and used it for the first time, notifications in the foreground worked. However, when I restarted the app, they stopped working.     

Realized the only difference was that I was calling `_firebaseMessaging.getToken();` to pass the token to the backend. 

So, getting firebase instance and calling `getToken` just before `FirebaseMessaging.onMessage.listen` worked for me.

```dart
final FirebaseMessaging _firebaseMessaging = FirebaseMessaging.instance;
_firebaseMessaging.getToken();
irebaseMessaging.onMessage.listen((RemoteMessage message) {
    print("message received in foreground");
  });
```
